Fix metric loss with re-scheduled checks

vickenty opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Since delaying sampler destruction for cancelled checks, checks that were quickly re-scheduled were getting errors from GetSender because the sampler for the check was still in the aggregator. A bit later, once deregisterSampler message is handled and the sampler is removed during flush, metrics from such checks would be completely lost.

This patch fixes the issue by making sampler re-use a normal condition instead of an error: now that the sampler can outlive the check (this is because we want to send the metrics from the previous run still waiting for the flush) it isn't an error to have a sampler exist when the check is scheduled.

To prevent race condition when registerSender may be called before deregisterSampler message is even processed by the aggregator, sender creation (or reuse) is now handled by a dedicated message as well. As long as registerSender is called before the check run, sampler will be created before metrics arrive: both sampler creation message and metrics are sent via the same channel.
### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Make an empty directory and place a openmetrics check configuration file inside:
```
mkdir -p tmp && cat > tmp/openmetrics.yaml <<EOF
init_config:
instances:
  - metrics: [ '*' ]
    prometheus_url: http://localhost:5000/telemetry
    namespace: check.metric
    service: qa
EOF
```

Run the agent with the following settings:

```
docker run --rm --name qa \
  -e DD_API_KEY=none \
  -e DD_TELEMETRY_ENABLED=true \
  -e DD_LOG_LEVEL=debug \
  -e DD_LOG_PAYLOADS=true \
  -e DD_HOSTNAME=qa \
  -e DD_AUTOCONF_CONFIG_FILES_POLL=true \
  -e DD_AUTOCONF_CONFIG_FILES_POLL_INTERVAL=2 \
  -v $(pwd)/tmp:/etc/datadog-agent/conf.d/openmetrics.d/ \
  IMAGE \
 | grep check.metric
```

Verify that the payloads from the openmetrics are present during flush, `service:qa` tag is present.

Trigger check re-scheduling, this needs to happen between flushes:

```
mv tmp/openmetrics.yaml{,.disabled}; sleep 5; mv tmp/openmetrics.yaml{.disabled,}
```

Check for at least 2-3 flushes that check continues to send metrics, and the `service:qa` tag is present. There should be no errors about missing sampler.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

vickenty committed:
Fix metric loss with re-scheduled checks

Since delaying sampler destruction for cancelled checks, checks that
were quickly re-scheduled were getting errors from GetSender because
the sampler for the check was still in the aggregator. A bit later,
once deregisterSampler message is handled and the sampler is removed
during flush, metrics from such checks would be completely lost.

This patch fixes the issue by making sampler re-use a normal condition
instead of an error: now that the sampler can outlive the check (this
is because we want to send the metrics from the previous run still
waiting for the flush) it isn't an error to have a sampler exist when
the check is scheduled.

To prevent race condition when registerSender may be called before
deregisterSampler message is even processed by the aggregator, sender
creation (or reuse) is now handled by a dedicated message as well. As
long as registerSender is called before the check run, sampler will
be created before metrics arrive: both sampler creation message and
metrics are sent via the same channel.

vickenty committed:
Adjust tests for delayed sampler creation

vickenty committed:
Separate test cases to avoid data races on the globals

Tests that re-init the demultiplexer conflict with the tests that do
not, because background goroutines for checks access the demultiplexer
instance when the next test starts.

This isn't an issue during normal agent run, because then the global
demultiplexer instance is read-only and not modified past initialization.

remeh reviewed (approved):
Elegant fix, LGTM!

I was wondering if there are benefits to "immediately" processing the registering stuff in the big `select` of the aggregator. My thoughts are that registering/deregistering queues could be processed after every flush, avoiding having to think about the concurrency of it, obviously at the cost of missing a collection window since we have to wait for the next Flush.
This said, current implementation with your fix seems perfectly fine to me already 👍

vickenty commented:
> I was wondering if there are benefits to "immediately" processing the registering stuff in the big select of the aggregator.
Registration events need to be synchronized with metrics:

vickenty commented:
> I was wondering if there are benefits to "immediately" processing the registering stuff in the big select of the aggregator. My thoughts are that registering/deregistering queues could be processed after every flush, avoiding having to think about the concurrency of it, obviously at the cost of missing a collection window since we have to wait for the next Flush.

The important bit there isn't that it happens immediately, but that it happens in specific order with the check metrics. Sending (de-)registration events on the same channel as metrics makes sure that the sampler is ready before we get the first metrics from the check, and we only remove the sampler once the final "Commit" from the check was handled.

vickenty committed:
Stop the collector before demux

Prevent the race in the tests where test workers access global demux
instance at the same time as demux.Stop clears it.
